docs/design/embedded-agent-worker.md

```
@@ -0,0 +1,597 @@
+  In-memory only: any live agent process was spawned by the live server (a server restart kills orphans and re-spawns workers with fresh tokens), so tokens never need to survive a restart; a stale token from a kill-escaped process is correctly rejected.
+- `/mcp` route (`mcp-server.ts:1570`): read `Authorization: Bearer <token>`, resolve via the registry, and expose the identity to tool handlers through an `AsyncLocalStorage<McpCallerIdentity | null>` wrapped around `transport.handleRequest(c)` (the MCP SDK does not thread HTTP context into handlers; ALS is the seam).
+- Enforcement helper used by tool handlers that accept a session-identity argument:
+
+```ts
+// mode from config: AGENT_CONSOLE_MCP_AUTH = 'off' | 'warn' | 'enforce'   (v1 default: 'warn')
+function checkCallerOwnsSession(caller: McpCallerIdentity | null, claimedSessionId: string, mode: Mode): ErrorResult | null
+```
+
+  Rules: (1) if a token WAS presented and the claimed session's `createdBy` differs from `caller.userId`, reject regardless of mode (a presented-but-mismatched identity is always an error, never a warning); (2) if no token was presented, `warn` logs and proceeds (today's behavior), `enforce` rejects; (3) `off` preserves today's behavior entirely. Apply at the elevation-bearing tools first (`delegate_to_worktree`, `remove_worktree`, `run_process`, `create_conditional_wakeup`) plus `send_session_message`'s `fromSessionId`.
+- **Token delivery, embedded-agent:** inside the stdin `init` message — never argv, never env. This is why the loop path can flip to hard enforcement immediately: activation always delivers a token.
+- **Token delivery, terminal agents (specified here, enforcement stays `warn` in v1):** non-elevated spawns may use a spawn env var (`AGENT_CONSOLE_MCP_TOKEN`) — env is not argv-visible on the direct path. Elevated spawns MUST NOT route the token through `buildElevationArgs` env embedding (it lands in the inner `sh -c` argv, world-readable via `/proc`); instead prepend `export AGENT_CONSOLE_MCP_TOKEN='<token>' && ` to the sentinel-injected agent command (the `pty.write` injection path — visible only inside the user's own terminal stream). How the terminal agent's MCP client attaches the header (`.mcp.json` header env expansion) must be verified during implementation; flipping the default to `enforce` is gated on that delivery landing.
```

markdownlint-cli2 (0.22.1)

[warning] Spaces inside code span elements

(MD038, no-space-in-code)